gson/src/main/java/com/google/gson/internal/bind/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,480 @@
+            }
+        }
+        Class<?> getType(Object recordComponent) {
+            try {
+                return (Class<?>) getType.invoke(recordComponent);
+            }
+            catch (IllegalAccessException | InvocationTargetException e) {
+                throw new TypeAdapterReflectionException(
+                        "Failed to invoke method [getGenericType] on RecordComponent", e);
+            }
+        }
+        Type getGenericType(Object recordComponent) {
+            try {
+                return (Type) getGenericType.invoke(recordComponent);
+            }
+            catch (IllegalAccessException | InvocationTargetException e) {
+                throw new TypeAdapterReflectionException(
+                        "Failed to invoke method [getGenericType] on RecordComponent", e);
+            }
+        }
+        // getAccessor returns a method, that in turn can be invoked on the Record to read out its value
+        Method getAccessor(Object recordComponent) {
+            try {
+                return (Method) getAccessor.invoke(recordComponent);
+            }
+            catch (IllegalAccessException | InvocationTargetException e) {
+                throw new TypeAdapterReflectionException("Failed to invoke method [getAccessor] on RecordComponent", e);
+            }
+        }
+        <A extends Annotation> A getAnnotation(Object recordComponent, Class<A> annotation) {
+            try {
+                return (A) getAnnotation.invoke(recordComponent, annotation);
+            }
+            catch (IllegalAccessException | InvocationTargetException e) {
+                throw new TypeAdapterReflectionException(
+                        "Failed to invoke method [getAnnotation] on RecordComponent", e);
+            }
+        }
+
+
+        public <T> Field getField(Object recordComponent, TypeToken<T> type) {
+            // There is nothing in the RecordComponent class to access the underlying field, so we do this based
+            // on the component name.
+            String fieldName = getName(recordComponent);
+            try {
+                return type.getRawType().getDeclaredField(fieldName);
+            }
+            catch (NoSuchFieldException e) {
+                throw new TypeAdapterReflectionException(
+                        "Expected to find field [" + fieldName + "] on recordComponent [" + recordComponent + "]"
+                        + " on type [" + type + "]. Somehow there is a discrepancy between the record components"
+                        + " and the fields on the Class.", e);
+            }
+        }
+
+    }
+
+    @Override
+    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
+        // ?: Is this not a record?
+        if (!RECORD_HELPER.isRecord(type)) {
+            // Yes -> Not supported, we return null as per the TypeAdapterFactory contract
+            return null;
+        }
+
```

Comment:
Probably this should also check the `ReflectionAccessFilter`:
https://github.com/google/gson/blob/2154e468b490c8a6576db13eca31c9d1afe6bf78/gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java#L104-L108

For example a user might only permit reflection based serialization of classes in a `model` package but disallow (accidental) serialization of classes in an `implementation` package.